### ISSUE TITLE
Unity object serialization

### ISSUE THREAD

**Issue by Eddio0141** [Eddio0141/UniTAS#201]

**Comment by sameerstg** [comment]
You can serialize the objects using [SerializedField] and class by [System.Serializable]

**Comment by Eddio0141** [comment]
Yeah I'm aware of it. I could apply those attributes at runtime with preload patcher.

Problem with that though is if there's already a class with `Serializable` on a game's type, and if one of the fields has `NonSerialized` on it, I could remove that attribute from the field but it could cause problems on the game's side down the line and we wouldn't know, while keeping that attribute won't be any good either for use on UniTAS side.

**Comment by Eddio0141** [comment]
Also for this issue, I mainly have to figure out serializing unity types (stuff like RigidBody, Texture2D, GameObject) and unless I'm missing something, I don't think its possible to serialize most of these even on a modern unity version

**Comment by Eddio0141** [comment]
# Steps to serializing
1. get all `GameObject`, active or not (and only the root)
2. serialize data for the `GameObject` instance, and its components
3. repeat for child objects for each root

# Things I'm not sure about
- should I get all Object and serialize all that?
- or maybe I get all Object, store in some "to be processed" list, and when a static field references an unity object, I remove from list and process it

# Notes
- All of the data is pretty much references
- Unity's objects can be referenced in static fields and pretty much in c# side

**Comment by Eddio0141** [comment]
# Idea 2, trace object creation steps
- Traces operations such as "scene load", "instantiate", etc
- I could use this as a "last resort" to some things I can't easily track
- Some operations can cancel out the previous
  - `scene load 1 -> scene load 2` can be reduced to just `scene load 2`, which also removes any "child operations" such as object manipulation
